### PR TITLE
Updates to pixelbuf API - Addresses #884

### PR DESCRIPTION
**Pull request by rhooper**
This work-in-progress still requires some thorough testing.  It switches over _pixelbuf to the same string based API as pypixelbuf.

Early feedback would be highly welcome, @dhalbert @tannewt 

Also thoughts on what a good test suite looks like would help me.

### CODE REVIEW

File: shared-bindings/_pixelbuf/PixelBuf.c [hunk 1]
```diff
@@ -87,23 +83,61 @@ STATIC mp_obj_t pixelbuf_pixelbuf_make_new(const mp_obj_type_t *type, size_t n_a
...
+    int num_chars = (dotstar ? 1 : 0) + (w ? 1 : 0) + (r ? 1 : 0) + (g ? 1 : 0) + (b ? 1 : 0);
+    if ((num_chars < byteorder_details.bpp) || !(r && b && g))
+        mp_raise_ValueError(translate("Invalid byteorder string"));
+    byteorder_details.is_dotstar = dotstar ? true : false;
+    byteorder_details.has_white = w ? true : false;
+    byteorder_details.byteorder.r = r - byteorder;
+    byteorder_details.byteorder.g = g - byteorder;
+    byteorder_details.byteorder.b = b - byteorder;
+    byteorder_details.byteorder.w = w ? w - byteorder : 0;
+    // The dotstar brightness byte is always first (as it goes with the pixel start bits)
```
Comment (tannewt) [inline shared-bindings/_pixelbuf/PixelBuf.c]: I'd suggest having D or `P` for pulse at the start so all characters represent a byte in order. It's weird to have a suffix shift everything.

File: shared-bindings/_pixelbuf/PixelBuf.c [hunk 2]
```diff
@@ -87,23 +83,61 @@ STATIC mp_obj_t pixelbuf_pixelbuf_make_new(const mp_obj_type_t *type, size_t n_a
...
+        mp_raise_TypeError(translate("byteorder is not a string"));
+
+    if (args[ARG_byteorder].u_obj == NULL)
+        byteorder = "BGR";
+    else 
+        byteorder = mp_obj_str_get_data(args[ARG_byteorder].u_obj, &bo_len);
+
+    if (bo_len < 3 || bo_len > 4)
+        mp_raise_ValueError(translate("Invalid byteorder string"));
+    strncpy(byteorder_details.order, byteorder, sizeof(byteorder_details.order));
```
Comment (tannewt) [inline shared-bindings/_pixelbuf/PixelBuf.c]: Why not hang onto the original object? Strings are immutable so it won't change.

File: shared-bindings/_pixelbuf/__init__.c [hunk 3]
```diff
@@ -167,159 +82,32 @@ const int32_t colorwheel(float pos) {
...
-PIXELBUF_BYTEORDER(LBRG, 4, 3, 1, 2, 0, false, true)
-//| .. data:: LBGR
-//|
-//|    * **order** *Luminosity*, Blue, Green, Red
-//|    * **bpp** 4
-//|    * **has_luminosity** True
-//|
-//|    Actual format commonly used by DotStar (5 bit luninance value)
-PIXELBUF_BYTEORDER(LBGR, 4, 3, 2, 1, 0, false, true)
-
```
Comment (tannewt) [inline shared-bindings/_pixelbuf/__init__.c]: Nice!

File: shared-bindings/_pixelbuf/PixelBuf.c (hunk 1 above)
Comment (rhooper) [reply shared-bindings/_pixelbuf/PixelBuf.c]: I originally had them be D prefixes (but P could work), but it just looked wrong.  I'm happy to make that change though, as it simplifies the code.

File: shared-bindings/_pixelbuf/__init__.c [hunk 4]
```diff
@@ -167,159 +82,32 @@ const int32_t colorwheel(float pos) {
...
-    { MP_ROM_QSTR(MP_QSTR_GBRW), MP_ROM_PTR(&byteorder_GBRW) },
-    { MP_ROM_QSTR(MP_QSTR_BRGW), MP_ROM_PTR(&byteorder_BRGW) },
-    { MP_ROM_QSTR(MP_QSTR_BGRW), MP_ROM_PTR(&byteorder_BGRW) },
-    { MP_ROM_QSTR(MP_QSTR_LRGB), MP_ROM_PTR(&byteorder_LRGB) },
-    { MP_ROM_QSTR(MP_QSTR_LRBG), MP_ROM_PTR(&byteorder_LRBG) },
-    { MP_ROM_QSTR(MP_QSTR_LGRB), MP_ROM_PTR(&byteorder_LGRB) },
-    { MP_ROM_QSTR(MP_QSTR_LGBR), MP_ROM_PTR(&byteorder_LGBR) },
-    { MP_ROM_QSTR(MP_QSTR_LBRG), MP_ROM_PTR(&byteorder_LBRG) },
-    { MP_ROM_QSTR(MP_QSTR_LBGR), MP_ROM_PTR(&byteorder_LBGR) },
+    { MP_ROM_QSTR(MP_QSTR_RGB),   MP_ROM_QSTR(MP_QSTR_RGB) },
```
[the remark this answers is not in the document]
Comment (rhooper) [reply shared-bindings/_pixelbuf/__init__.c]: OK - These have been removed.

**Comment by rhooper** [comment]
@dhalbert @tannewt 
I'm not all that happy with changing the signature for subscr globally, but couldn't find a non-breaking way to ensure native subscript methods are able to call instance methods.

The issue was that mp_obj_subscr calls instance_subscr that then called mp_obj_subscr with self->subobj[0] (which is the native type).  The native subscr could no longer call the non-native subclass method because the argument to pixelbuf_pixelbuf_subscr was the native class and not the NeoPixel subclass of it.

I tried some other tricks, such as changing the calling convention for instance_subscr, but ultimately broke some unit tests.

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -15,7 +15,10 @@ jobs:
     - name: Dump GitHub context
       env:
         GITHUB_CONTEXT: ${{ toJson(github) }}
-      run: echo "$GITHUB_CONTEXT"
+      run: python -c "import os; print os.environ['GITHUB_CONTEXT']"
+    - name: Fail if not a release publish # workaround has `on` doesn't have this filter
+      run: exit 1
+      if: github.event_name == 'release' && (github.event.action != 'published' && github.event.action != 'rerequested')
```
Comment (tannewt) [inline .github/workflows/build.yml]: This shouldn't be added. It's hack we no longer need.

File: py/objtype.c [hunk 6]
```diff
@@ -852,7 +852,7 @@ STATIC mp_obj_t instance_subscr(mp_obj_t self_in, mp_obj_t index, mp_obj_t value
         meth_args = 3;
     }
     if (member[0] == MP_OBJ_SENTINEL) {
-        return mp_obj_subscr(self->subobj[0], index, value);
+        return mp_obj_subscr_impl(self->subobj[0], index, value, instance);
```
Comment (tannewt) [inline py/objtype.c]: Did you try passing `self` here instead of `self->subobj[0]`? You'll need to add it to all of the native calls where you modified method signature. That will ensure that all native subscrs receive the actual object (native or not).

Comment (rhooper) [reply py/objtype.c]: I had planned to do this - now done.

Comment (tannewt) [reply py/objtype.c]: These changes aren't what I meant. I meant leaving the signature as it was originally and passing self here rather than the native "subobject". To do so, each native subscr will have to get the native object itself instead. (Except for yours which will use both.)

Comment (rhooper) [reply py/objtype.c]: I'll have a look around and see if that approach would work.  I feel like I tried that and ended up in an infinite loop because the calling path is:

mp_obj_subscr => instance_subscr => mp_obj_subscr
(because of o->subscr = instance_subscr in mp_obj_new_type)

Comment (tannewt) [reply py/objtype.c]: Ah right! Seems to me that instance_subscr should be smarter itself rather than calling back to mp_obj_subscr.

Comment (rhooper) [reply py/objtype.c]: If you can think of how to make instance_subscr smarter, i'd love to know your thoughts.  Ultimately it does that when the base is native and has a subscr (see mp_obj_class_lookup ~ line 168) `lookup->dest[0] = MP_OBJ_SENTINEL;`

I guess we could make it look up and call the native subscr method itself, but then all of the assertions mp_obj_subscr could throw need reimplementing.

Comment (tannewt) [reply py/objtype.c]: So, I think the native call needs to be moved from mp_obj_subscr to obj_instance. As is, a subclass of a native object won't be able to override subscr. This won't require reimplementing or moving the errors thrown by mp_obj_subscr because instance_subscr can return MP_OBJ_NULL to prevent the early return and cause an error.

The native call also needs to pass in the original instance and each native implementation will need to get the subobj to use. That will allow the native subscr full access to the instance.

Comment (rhooper) [reply py/objtype.c]: If I understand you correctly, instead of instance_subscr calling mp_obj_subscr, it should do something like this:

```
    if (member[0] == MP_OBJ_SENTINEL) {
        //return mp_obj_subscr(self->subobj[0], index, value, instance);

        mp_obj_type_t *type = mp_obj_get_type(self->subobj[0]);
        if (type->subscr != NULL) {
            mp_obj_t ret = type->subscr(self_in, index, value, instance);
            // May have called port specific C code. Make sure it didn't mess up the heap.
            assert_heap_ok();
            if (ret != MP_OBJ_NULL) {
                return ret;
            }
        }
        return MP_OBJ_NULL;
```

This fails the unit tests.  Much like many of my first attempts to make this work.

> 2 tests failed: class_store_class subclass_native1

### Expected
b'[1, 2, 5]\nsomething\n5\n[-1, 2, 5]\n3\n[-1, 2, 5, 20, 30, 40]\nTypeError\nTypeError\n'
### Actual
b'[1, 2, 5]\nsomething\nsomething\n[1, 2, 5]\n3\n[1, 2, 5, 20, 30, 40]\nTypeError\nTypeError\n'
FAIL  basics/subclass_native1.py
### Expected
b"b'a' b'b'\nAll's ok\n"
### Actual
b'CRASH'
FAIL  basics/class_store_class.py
2 tests performed (10 individual testcases)
0 tests passed
2 tests failed: class_store_class subclass_native1

Comment (tannewt) [reply py/objtype.c]: Did you update list's subscr as well? It'll need to get subobj[0] itself. Not doing so could explain the first failure.

The second will fail if namedtuple's native object subscr hasn't been updated.

Comment (rhooper) [reply py/objtype.c]: I did not update any subscr methods.  I'll give those two a try, and if that works, I'll know what to check for codebase-wide.

Comment (rhooper) [reply py/objtype.c]: @tannewt while working on this approach, I was partly successful. 
In instance_subscr i fetched `subobj[0]` to get the type, then called `subobj_type->subscr(self_in, index, value).`

For some of the _subscr implementations I used `mp_instance_cast_to_native_base`, for example, in `list_subscr` which appeared to work, and satisfied the test suite... But when it came to `bytes_subscr` which is used by both `str` and `bytes`, I couldn't find a straightforward solution that didn't break the test suite (or crash).

Comment (tannewt) [reply py/objtype.c]: Getting the subobj directly in bytes_subscr didn't work? Wouldn't that be the equivalent of what it's doing now?

Comment (rhooper) [reply py/objtype.c]: I ended up with crashes.  I'm wary of significantly modifying the code here without fully understanding it.  
https://github.com/adafruit/circuitpython/blob/master/py/objstr.c#L443

Comment (tannewt) [reply py/objtype.c]: Is the change anywhere? I'd be happy to poke at it. Was it tests that were crashing?

Comment (rhooper) [reply py/objtype.c]: @tannewt https://github.com/rhooper/circuitpython/tree/pixelbuf-subscr-change is the WIP

Comment (tannewt) [reply py/objtype.c]: So the crash thing is not as bad as it sounds. The class_store_class test was just throwing an exception that subscr wasn't supported. I've fixed it here: https://github.com/rhooper/circuitpython/pull/2

I don't think there is much you need to change actually. Most of the native subscr methods are called directly rather than through instance_subscr. instance_subscr only comes into play with native class when they are subclassed. Most native objects don't work correctly when subclassed so that isn't generally an issue. namedtuples are a very special case where they are subclassable and they themselves pretend to be a subclass of tuple.

**Comment by ladyada** [comment]
@rhooper hi - what are ya currently blocking on for this PR - we can get some help to wrap up this PR.

**Comment by rhooper** [comment]
@ladyada 

From the top of my head, these are the remaining tasks:

- [ ] Free up time somehow!
- [ ] Sort out making it so native method calls can call python subclass overrides (specifically so the native `.fill()` can call `.show()` on the python subclass).  This is needed for `auto_write` to work properly with the native `.fill()`.  We could skip native .fill() for now.
- [ ] Finish writing pypixelbuf so that implementations without pixelbuf work
- [ ] Test and merge the neopixel changes that work with pixelbuf (and pypixelbuf)
- [ ] Test and merge adafruit_dotstar changes that work with pixelbuf (and pypixelbuf)
- [ ] Test Test Test
- [ ] Update which neopixel and dotstar libs are frozen in

**Comment by ladyada** [comment]
@rhooper  would it be ok for other folks to implement/test some of these TODOs?

**Comment by rhooper** [comment]
@ladyada absolutely fine with me. the more people that know the changes the better! 
all i’d ask is that work be coordinated so i don’t spend time doing something someone else is working on.

**Comment by ladyada** [comment]
sure, they'll post here in this thread

**Comment by rhooper** [comment]
I worked around the native `.fill()` method being unable to call `.show()` on subclasses using a helper function in _pixelbuf instead.

**Comment by rhooper** [comment]
Dotstar and Neopixel changes appear to work, but I need some assistance gathering up examples to test compatibility.

https://github.com/adafruit/Adafruit_CircuitPython_NeoPixel/pull/59
https://github.com/adafruit/Adafruit_CircuitPython_DotStar/pull/42

**Comment by rhooper** [comment]
Pypixelbuf has had some testing done - seems to work - but needs the same kind of testing as the Dotstar and NeoPixel PR.  It also needs putting in pypi so that the Dotstar and Neopixel builds can pass.

https://github.com/adafruit/Adafruit_CircuitPython_Pypixelbuf/pull/3

File: extmod/modbtree.c [hunk 7]
```diff
@@ -246,8 +246,10 @@ STATIC mp_obj_t btree_iternext(mp_obj_t self_in) {
     }
 }
 
+#pragma GCC diagnostic ignored "-Wunused-parameter"
```
Comment (tannewt) [inline extmod/modbtree.c]: Why is this needed? Does it apply to the whole file? I've done `(void) var;` instead. That is clearer about which parameter is unused.

File: shared-bindings/_pixelbuf/PixelBuf.c [hunk 8]
```diff
@@ -349,23 +332,14 @@ STATIC mp_obj_t pixelbuf_pixelbuf_unary_op(mp_unary_op_t op, mp_obj_t self_in) {
 
 //|   .. method:: show()
 //|
-//|     Call the associated write function to display the pixels.
+//|     Does nothing unless subclassed.
```
Comment (tannewt) [inline shared-bindings/_pixelbuf/PixelBuf.c]: Think we should raise a `NotImplementedError`?

File: shared-bindings/_pixelbuf/PixelBuf.h [hunk 9]
```diff
@@ -36,21 +36,18 @@ typedef struct {
...
     uint8_t *rawbuf;
     uint8_t *buf;
-    mp_obj_t write_function;
-    mp_obj_tuple_t *write_function_args;
     bool auto_write;
 } pixelbuf_pixelbuf_obj_t;
 
 void pixelbuf_recalculate_brightness(pixelbuf_pixelbuf_obj_t *self);
-void call_write_function(pixelbuf_pixelbuf_obj_t *self);
+mp_obj_t call_show(mp_obj_t self_in);
```
Comment (tannewt) [inline shared-bindings/_pixelbuf/PixelBuf.h]: I'd prefix this with `pixelbuf_` for clarity.

File: shared-bindings/_pixelbuf/PixelBuf.c (hunk 8 above)
Comment (rhooper) [reply shared-bindings/_pixelbuf/PixelBuf.c]: Should we add another error string, or is `mp_raise_NotImplementedError(NULL)` ok here?

File: extmod/modbtree.c (hunk 7 above)
Comment (rhooper) [reply extmod/modbtree.c]: This is leftover from the `_subscr` changes - removing these.

File: shared-bindings/_pixelbuf/PixelBuf.h (hunk 9 above)
Comment (rhooper) [reply shared-bindings/_pixelbuf/PixelBuf.h]: Good call.

File: shared-bindings/_pixelbuf/PixelBuf.c (hunk 8 above)
Comment (tannewt) [reply shared-bindings/_pixelbuf/PixelBuf.c]: NULL is fine.

**Comment by ladyada** [comment]
💯 💯 🔥 🏅 💯 💯  to everyone who helped get this over the finish line - you are heros! many glowy projects will thank u :)

**Comment by siddacious** [comment]
:🎉🎉🎉Whoooo🎉🎉🎉🥳

Awesome contribution, thank you! Much glow will be had
